lib/husky_musher/utils/redcap.py

```diff
@@ -507,30 +510,40 @@ def kiosk_registration_link(redcap_record: dict, instances: Dict[str, int]) -> s
     appropriate survey link to the correct instance of a Kiosk Registration
     instrument according to the pre-determined logic flow.
     """
-    record_id = redcap_record['record_id']
-    event = 'encounter_arm_1'
-    instrument = 'kiosk_registration_4c7f'
     incomplete_kr_instance = instances['incomplete_kr']
 
     if need_to_create_new_td_for_today(instances):
         # Create TD instance based on # of days since project start.
         create_new_testing_determination(redcap_record)
-
-        # Generate a link to the KR, and then redirect.
-        survey_link = generate_survey_link(record_id, event, instrument,
-            get_todays_repeat_instance())
+        instance = get_todays_repeat_instance()
 
     elif need_to_create_new_kr_instance(instances):
-        # Generate a link to the target instance, and then redirect.
-        survey_link = generate_survey_link(record_id, event, instrument,
-            instances['target'])
+        instance = instances['target']
 
     elif incomplete_kr_instance is not None:
-        # Generate a link to the existing KR that is incomplete, and then redirect.
-        survey_link = generate_survey_link(record_id, event, instrument,
-            incomplete_kr_instance)
+        instance = incomplete_kr_instance
 
     else:
         raise Exception("Logic error when generating survey links.")
 
-    return survey_link
+    return generate_redcap_link(redcap_record, instance)
+
+
+def generate_redcap_link(redcap_record: dict, instance: int):
+    """
+    Given a *redcap_record*, generate a link to the internal REDCap portal's
+    Kiosk Registration form for the record's given REDCap repeat *instance*.
+    """
+    project = Project(REDCAP_API_URL, REDCAP_API_TOKEN, PROJECT_ID)
```

Comment:
Constructing a `Project` every time will cause a request to REDCap to verify project id and token match. It would be faster to only construct a `Project` at app start instead of every time this function is called. (This will also make the object available for other functions to use if we want to port them to use it instead of `requests`.)

Reply:
Thanks. I should have thought about this more after realizing how much creating a `Project` slowed down my unit tests. I'll push a refactor. Thanks!

Reply:
@tsibley do you have any thoughts on constructing `Project` within `app.py` and passing it to the `kiosk_registration_survey()` function vs. creating a global variable `PROJECT` in `redcap.py` and allowing the functions there to access it from the scope?

Reply:
Hmm. @kairstenfay Both of your suggestions should reduce instantiations of the Project compared to current state. But my understanding is that most methods of making a global variable will still get constructed again for every new route or request still - @tsibley may be able to better speak to how to create/store the object so it only gets created at app start?

Reply:
> But my understanding is that most methods of making a global variable will still get constructed again for every new route or request still

Which ways are you thinking of?

A global variable (at the file/module level) should only get constructed once per process at module load time, e.g.:

https://github.com/seattleflu/husky-musher/blob/6405835a328b61338f802e1532f88fdad0846609/lib/husky_musher/utils/redcap.py#L10-L12

Since there are side effects (network requests) to constructing a `Project`, we might want make construction lazy, e.g. by delaying until first use:

https://github.com/seattleflu/id3c/blob/697c8cf5943bb06af55d91d7df9fe64426cd8a4c/lib/id3c/cli/command/geocode.py#L431-L433

Reply:
I pushed a new commit defining `Project` as a global variable in redcap.py. Please take a look when you get a chance. Note that I could not get unit tests to run when constructing `Project` lazily.

Reply:
Aha, I see. I was thinking of using flask's `g` or other mechanisms to store in the application context , (https://flask.palletsprojects.com/en/1.1.x/appcontext/#storing-data) which come into play when you need to persist data or a variable across a request, but don't want it to be shared across different threads which may modify the data in an unexpected manner or order. We don't have the constraint of worrying about different threads modifying `PROJECT`, because we interact with it in a read only way. Putting it as a global in the module seems to work. Thanks for helping me think through this 👍